### PR TITLE
Refac user login

### PR DESCRIPTION
**Pull request by miltonbsn**
# Summary 

Since we are upgrading most libs, including express and node.js itself, I am proposing a major refactoring to follow the good practices introduced since I developed the first version of these features. Besides apply new es6 standard code, I am changing significantly the core architecture on backend, following a domain driven design standard. The main goal is to allow new features to be easily introduced and tested

## List of changes

- Separates the coupling between app and server ;
- Complete rewrite of user endpoints, introducing a new package called user with all it's responsibilities; 
- Add tests with Jest, supertest and mockingoose;

### CODE REVIEW

**Comment by miltonbsn** [comment]
@feekosta can I have your first thoughts here?

File: server_app/user/service.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+const login = async ({username, password, sessionId}) => {
+  return new Promise(async (resolve, reject) => {
+    try {
+      const userDocument = await UserRepository.findOne({
+        "login": username,
+        "password": encriptor.Crypto(password, username),
+      });
+    
+      if(userDocument != null) {
+        resolve({
```
Comment (feekosta) [inline server_app/user/service.js]: return in the if, to not execute the rest of the code

File: server_app/user/service.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  });
+}
+
+const create = async ({username, password, mail}) => {
+  return new Promise(async (resolve, reject) => {
+    try {
+      const user = await UserRepository.findOne({"login": mail});
+
+      if(user != null) {
+        reject({"code": "USER_ERROR_ALREADY_EXISTS"});
```
Comment (feekosta) [inline server_app/user/service.js]: return in the if, to not execute the rest of the code

**Review by feekosta — CHANGES_REQUESTED** [review]

**Comment by miltonbsn** [comment]
@rpuggian you may be interested on this

**Review by feekosta — APPROVED** [review]

File: server_app/user/service.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  return new Promise(async (resolve, reject) => {
+    try {
+      const userDocument = await UserRepository.findOne({
+        "login": username,
+        "password": encriptor.Crypto(password, username),
+      });
+    
+      if(userDocument != null) {
+        return resolve({
+          "sessionId": sessionId,
```
Comment (rpuggian) [inline server_app/user/service.js]: Miltão, What do you think about leaving the responsibility of setting up the user session to the handler? I noticed in the handler we already have the session ID. Maybe like that we don't need to verify != null here and always resolve the userDocument.

File: server_app/user/handler.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    const password = req.body.password;
+    const sessionId = req.sessionID;
+
+    const validation = userValitor.validateLoginParams({username, password});
+
+    if(!validation.valid) {
+      return res.status(422).send(validation.message);
+    }
+      
+    const userSession = await userService.login({username, password, sessionId});
```
Comment (rpuggian) [inline server_app/user/handler.js]: Maybe you can build the user session here, and just find the user in the userService using the password encryption.

**Comment by idmarjr** [comment]
Folks, are we good to go?

**Pull request merged**